frontend/document/documentregistry.lua

```diff
@@ -9,22 +9,47 @@ local DocumentRegistry = {
     providers = {},
 }
 
-function DocumentRegistry:addProvider(extension, mimetype, provider)
-    table.insert(self.providers, { extension = extension, mimetype = mimetype, provider = provider })
+function DocumentRegistry:addProvider(extension, mimetype, provider, weight)
+    table.insert(self.providers, {
+        extension = extension,
+        mimetype = mimetype,
+        provider = provider,
+        weight = weight or 100,
+    })
 end
 
---- Returns the registered document handler.
+--- Returns the preferred registered document handler.
 -- @string file
 -- @treturn string provider, or nil
 function DocumentRegistry:getProvider(file)
+    local providers = self:getProviders(file)
+
+    if #providers >= 1 then
+        return providers[1].provider
+    end
+end
+
+--- Returns the registered document handlers.
+-- @string file
+-- @treturn string provider, or nil
+function DocumentRegistry:getProviders(file)
+    local providers = {}
+
     -- TODO: some implementation based on mime types?
     for _, provider in ipairs(self.providers) do
         local suffix = string.sub(file, -string.len(provider.extension) - 1)
         if string.lower(suffix) == "."..provider.extension then
         -- if extension == provider.extension then
-            return provider.provider
+            -- stick highest weighted provider at the front
+            if #providers >= 1 and provider.weight > providers[#providers].weight then
+                table.insert(providers, #providers, provider)
```

Comment:
Shouldn't the two `#providers` be `1` if you want to stick it at the front ?

Reply:
Definitely. Kind of a half-finished thought about sorting by weight when I realized that doesn't really matter I suppose.